### PR TITLE
feat(wallet) save/load collectibles user handled state in DB

### LINKED ISSUE #13313: [Tokens Management] Missing collectibles settings integration with backend
### Description

The group order saving is missing. I tried implementing the group order saving by building on the current generic sorting data structure but it seems to be some corner cases that don't work; see https://github.com/status-im/status-desktop/pull/13970
Therefore, I created a separated task for groups settings here: https://github.com/status-im/status-desktop/issues/13971

Currently the UI stores the settings created by the user in the `local settings` file.

This task pretends to cover the link between the UI and the existing NIM backend in order to store the settings data into the DB.

Maybe there are some missing parameters to take into account in the backend side about settings. Here a reminder:

- [Sorting order](https://www.figma.com/file/eM26pyHZUeAwMLviaS1KJn/%E2%9A%99%EF%B8%8F-Wallet-Settings%3A-Manage-Tokens?type=design&node-id=32%3A138356&mode=design&t=GTJPXvZ6JuQc6YkN-1): The DB needs to store the sorting type applied in order to provide the initial value to the UI. Whenever the user changes the ordering type, it must be saved in the DB.

Screenshot 2024-01-26 at 11 54 27

- Custom order must include:
  - [x] Individual collectible item position.
  - [x] Hidden state of individual collectible item.
- Backend shall provide 2 different models to the UI, the `flat collectibles list model` AND the `groups list model` (being this last one, the community groups or the collections groups with the relevant information about the community and how many elements this group contains).
- There shall be available a backend call to hide elements one by one, or a specific group.
- There shall be available a backend call to get and set the new elements position (being elements an individual item or a group item).

### PR DESCRIPTION
**Pull request by stefandunca**
### Closes #13313, #13971 and Updates #13312

- [x] rebase on top of master
- [x] fix collectibles type

Add a separation layer for save/load/clear to `ManageTokensModel` so that we can save/load from external sources.
The separate layer is composed of JSON as protocol, a set of signals and slots for interface. The implementation forwards data to current `QSettings` for storybook and nim controllers for the app.

Groups are not handled in this PR

Note: I auto formatted some C++ files by mistake using vscode, please use ignore spaces in diff for easier review :(

### CODE REVIEW

**Comment by stefandunca** [comment]
@caybro and @dlipicar I worked on this task of redirecting the collectible settings from being stored in `QSettings` by the `ManageTokensController` C++ implementation to be stored in the DB (by calling `updateCollectiblePreferences`).
However, it seems it is becoming quite complicated and error prone and the initial estimation was "easy".

I'm planning to emit signals for save/load events and call the load/save of `ManageTokensController` with json data. However, I see that the c++ controller is used also for assets which have a different structure and options so the JSON has to be flexible and detect the source based on presence of some properties.

Could you please have a look at the current code and see if there is an easier way to do that that I might have missed?

File: ui/StatusQ/src/wallet/tokendata.cpp
Comment (caybro) [inline ui/StatusQ/src/wallet/tokendata.cpp:56]: This should also handle "assets" as well... Otherwise the main wallet Assets view will not work

File: ui/StatusQ/src/wallet/tokendata.cpp
Comment (caybro) [inline ui/StatusQ/src/wallet/tokendata.cpp:58]: This is fixed in my PR here: https://github.com/status-im/status-desktop/pull/13984

**Review by caybro — CHANGES_REQUESTED** [review]
In general I like the direction where this is going however...

I don't think it's desirable to merge this as-is at this point in the release cycle (during a bugfixing period) when this is clearly:
- new feature
- breaking change
- incomplete (deals with collectibles only)

File: ui/StatusQ/src/wallet/tokendata.cpp
Comment (stefandunca) [reply ui/StatusQ/src/wallet/tokendata.cpp:56]: Assets work also but they go in QSettings  as was previously and not in DB as with collectibles. I see  no reason not to do it in multiple steps. It is already a too big of a change.

File: ui/StatusQ/src/wallet/tokendata.cpp
Comment (stefandunca) [reply ui/StatusQ/src/wallet/tokendata.cpp:58]: I will take into consideration when tackling assets next iteration

File: ui/StatusQ/src/wallet/tokendata.cpp
Comment (stefandunca) [reply ui/StatusQ/src/wallet/tokendata.cpp:56]: Also I don't plan to support migration from QSettings to DB so there is no good reason to push back changes unti all categories as finalized.

**Review by noeliaSD — COMMENTED** [review]
Really nice work, code LGTM!
Here my comments, there are just some questions or suggestions!

I cannot test the pr properly to ensure there are no regressions bc right now we have a bug on master that doesn't apply settings to the main wallet view. It should be solved in here: https://github.com/status-im/status-desktop/pull/14018 and it would be nice if you can rebase on top of that to be able to do some proper testing.

Then, related to the topic @caybro exposed, as I explained offline, we need to consider if this refactoring PR should be integrated on this estabilization phase or wait for the next milestone. Considering that the `2.28 rc branch` will be create this Wednesday, maybe we can keep this pr open until the branch is created and integrate later, so this code will go to the `2.29` release. WDYT? 

I don't really have strong opinion on it if we can do a full and proper test to ensure there are no regressions introduced and we also ensure the bug fixed in here https://github.com/status-im/status-desktop/pull/13984 is also integrated.

CC: @alexjba @micieslak

File: ui/StatusQ/src/wallet/tokendata.h
Comment (stefandunca) [reply ui/StatusQ/src/wallet/tokendata.h:10]: I just followed the original tuple by improving on it with the purpose to make changes in small steps. I will take this into consideration in a followup change

File: ui/StatusQ/src/wallet/tokendata.h
Comment (stefandunca) [reply ui/StatusQ/src/wallet/tokendata.h:24]: This i storing also assets. I just followed the original tuple by improving on it with the purpose to make changes in small steps. I will take this into consideration in a followup change

File: ui/StatusQ/src/wallet/tokendata.h
Comment (stefandunca) [reply ui/StatusQ/src/wallet/tokendata.h:10]: That abstraction doesn't help the current task much so I consider it should be left for a separate refactoring/redesign phase.

File: ui/app/AppLayouts/Wallet/panels/ManageAssetsPanel.qml
Comment (stefandunca) [reply ui/app/AppLayouts/Wallet/panels/ManageAssetsPanel.qml:25]: I agree with you. However, due to the current design we have the limitation of calling directly the Nim storing APIs. Therefore we have to move through the common layer, QML. Hence splitting the API into intention/pre/save/post actions from which the save has to be done in QML to have access to Nim asset/collectible controller.

File: ui/app/AppLayouts/Wallet/stores/CollectiblesStore.qml
Comment (stefandunca) [reply ui/app/AppLayouts/Wallet/stores/CollectiblesStore.qml:35]: I don't see an easier way. The save is external API (Nim) but start/finish is internally required by the `ManageTokensController` (C++) internal management. Therefore they have to go through the common medium `QML`.

**Review by stefandunca — COMMENTED** [review]
> I cannot test the pr properly to ensure there are no regressions bc right now we have a bug on master that doesn't apply settings to the main wallet view. It should be solved in here: #14018 and it would be nice if you can rebase on top of that to be able to do some proper testing.

I see a lot of parallel work on token master. I talked with you, @Cuteivist and reach out to @caybro and @dlipicar but still end-up now with parallel refactoring and major changes. Where can I find the work streams on token management (mainly `managetokenscontroller.cpp` and `managetokensmodel.h`) so that I could stream the rebase, review, regression test and not drag it forever?

> Then, related to the topic @caybro exposed, as I explained offline, we need to consider if this refactoring PR should be integrated on this estabilization phase or wait for the next milestone. Considering that the 2.28 rc branch will be create this Wednesday, maybe we can keep this pr open until the branch is created and integrate later, so this code will go to the 2.29 release. WDYT?

In my testing the functionality was kept and worked fine both in management view and display model. I trust your judgement and will follow with rebase and conflict after the parallel redesign for later on.

> I don't really have strong opinion on it if we can do a full and proper test to ensure there are no regressions introduced ...

Testing is waiting for a rebase, however it make no sense to rebase until we merge the all conflicting changes.

> and we also ensure the bug fixed in here https://github.com/status-im/status-desktop/pull/13984 is also integrated.

At first glance that change is implementing group ordering and conflicting with my changes. So I expect a rebase to be costly. As discussed privately, I will follow your collective decision and wait for all new features and fixes be in master, then rebase and adapt to the new changes.

**Comment by stefandunca** [comment]
@lukaszso I rebased on top of https://github.com/status-im/status-desktop/pull/14018

File: ui/app/AppLayouts/Wallet/stores/WalletAssetsStore.qml
Comment (micieslak) [inline ui/app/AppLayouts/Wallet/stores/WalletAssetsStore.qml:25]: why in this case this started/finished pair is not used?

File: ui/StatusQ/src/wallet/tokendata.h
Comment (noeliaSD) [reply ui/StatusQ/src/wallet/tokendata.h:10]: Sure, suggestion to take into account for a next step.

**Review by micieslak — APPROVED** [review]
Even though the code is still quite complicated I think it's a move into a right direction. `ManageTokensController` seems to have too many responsibilities inside and should be split into smaller chunks, excluding persistence layer is really good first step here.

I see some benefits if we could merge it asap - it may safe us from handling migrations from qsettings to db-base storage in the future.

At first glance allowing controller to decide when to load/save makes it more complex than it should be, maybe we could reconsider it in the future.

In the long term I think there is refactor needed, similar as the one done for Profile Showcase Management. `CollectiblesView` and `AssetsView` should not depend on the controller at all. The order/visibility info could be attached to the main model using LeftJoinModel making things much easier and cleaner. But I'm aware it requires deeper analysis bc of grouping and additional transformations needed bc of that. Probably `GroupingModel` should be finished first (https://github.com/status-im/status-desktop/issues/12683) to make it doable.

**Comment by stefandunca** [comment]
> @lukaszso I rebased on top of #14018

@noeliaSD tested and reported that the fix for apply changes doesn't work with my changes. I will investigate this issue next week when I pick it up again.

**Comment by alexjba** [comment]
> Then, related to the topic @caybro exposed, as I explained offline, we need to consider if this refactoring PR should be integrated on this estabilization phase or wait for the next milestone. Considering that the `2.28 rc branch` will be create this Wednesday, maybe we can keep this pr open until the branch is created and integrate later, so this code will go to the `2.29` release. WDYT?

I'd wait for the release branch on Wednesday before merging it, but no longer after that. It could land in 2.28.1 depending on how it will be branched out.

**Comment by anastasiyaig** [comment]
this task is for 2.28 so we need to merge it before release branch cut. Other words - this change is needed for 2.28

**Comment by stefandunca** [comment]
@micieslak I can't reply to your question so I answer in a comment

https://github.com/status-im/status-desktop/pull/13903#discussion_r1530224806
> Yeah, I also wonder why the emitter cannot call on his own this savingStarted/savingFinished pair?

The main reason is the inability to guarantee that a slot will answer to the save request, hence avoid unnecessary invalidation of the model (this way start and finish notify the controller of storage being active and mark the invalidation).

Another consideration was to be explicit and not to risk an inconstancy in case of connection type between `onRequestSaveSettings` and start/finish processing. However, now I see that this can still happen and is not guaranteed and your proposal would be as equally working given our simple integration (same thread affinity).

File: ui/app/AppLayouts/Wallet/stores/WalletAssetsStore.qml
Comment (stefandunca) [reply ui/app/AppLayouts/Wallet/stores/WalletAssetsStore.qml:25]: They are already done serialy in the internal `saveToQSettings` as was done before. The marker methods started/finished are just for external storage as discussed above.

**Comment by caybro** [comment]
> this task is for 2.28 so we need to merge it before release branch cut. Other words - this change is needed for 2.28

Why is it needed and what added value or fix does it bring to the user?

**Comment by anastasiyaig** [comment]
@caybro no idea, i just checked the milestone for the task :) @iurimatias @alexandraB99 please clarify

**Comment by stefandunca** [comment]
> At first glance allowing controller to decide when to load/save makes it more complex than it should be, maybe we could reconsider it in the future.

I totally agree, the implementation should be refactored with more separation of concerns in mind. I found it outside my task requirements and tried to do an initial quick implementation that escalated in too many long debugging sessions but good learnings.

**Comment by lukaszso** [comment]
Tested after rebase and the "Apply to my wallet" worked at first albeit with a delay (it looked like the changes were reflected only after the wallet activity was refreshed?). Then after about 30 minutes of inactivity it stopped working again and i am unable to get it to apply the order again.

Other than this i did not find any other issues here.

**Comment by noeliaSD** [comment]
I agree with @alexjba and @micieslak on landing this into `2.28.1`. One of the reasons to consider that way is (apart from the starting point of reducing complexity on the controller's data management):
- _Michal_ comment --> it may safe us from handling migrations from qsettings to db-base storage in the future. _Noelia_ comment --> Future that will come soon bc it's a product requirement afaik.

I don't see any risk if we can do a full testing and ensure the existing functionality is still working properly.

**Comment by stefandunca** [comment]
@lukaszso I just rebased on latest master and fixed all found issues. Can you please give it another try?

**Review by lukaszso — APPROVED** [review]
Tested and approved. I did not see any issues with assets/collectibles except for long wait for new ordering to be applied. It's not related to this change so i'll report it separately. LGTM :)

**Pull request merged**
